src/main/java/io/airlift/units/Count.java

```diff
@@ -0,0 +1,203 @@
+        if (scale * magnitude.getFactor() != this.magnitude.getFactor()) {
+            throw new IllegalArgumentException(format("Unable to represent %s in %s, conversion would cause a precision loss", this, magnitude));
+        }
+        try {
+            return multiplyExact(value, scale);
+        }
+        catch (ArithmeticException e) {
+            throw new IllegalArgumentException(format("Unable to represent %s in %s due the Long value overflow", this, magnitude));
+        }
+    }
+
+    public Count convertTo(Magnitude magnitude)
+    {
+        return new Count(getValue(magnitude), magnitude);
+    }
+
+    /**
+     * @return converted count with no bigger value than 1000 in succinct magnitude, fractional part is rounded
+     */
+    public Count convertToMostSuccinctRounded()
+    {
+        for (Magnitude magnitude : MAGNITUDES) {
+            double converted = (double) value * this.magnitude.getFactor() / magnitude.getFactor();
+            if (converted < 1000) {
+                return new Count(Math.round(converted), magnitude);
+            }
+        }
+        throw new IllegalStateException();
+    }
+
+    @JsonValue
+    @Override
+    public String toString()
+    {
+        return value + magnitude.getMagnitudeString();
+    }
+
+    @JsonCreator
+    public static Count valueOf(String count)
+            throws IllegalArgumentException
+    {
+        requireNonNull(count, "count is null");
+        checkArgument(!count.isEmpty(), "count is empty");
+
+        Matcher matcher = PATTERN.matcher(count);
+        if (!matcher.matches()) {
+            throw new IllegalArgumentException("Not a valid count string: " + count);
+        }
+
+        long value = parseLong(matcher.group(1));
+        String magnitutdeString = matcher.group(2);
+
+        for (Magnitude magnitude : Magnitude.values()) {
+            if (magnitude.getMagnitudeString().equals(magnitutdeString)) {
+                return new Count(value, magnitude);
+            }
+        }
+
+        throw new IllegalArgumentException("Unknown magnitude: " + magnitutdeString);
+    }
+
+    @Override
+    public int compareTo(Count o)
+    {
+        return Long.compare(getValue(Magnitude.SINGLE), o.getValue(Magnitude.SINGLE));
+    }
+
+    @Override
+    public boolean equals(Object o)
+    {
+        if (this == o) {
+            return true;
+        }
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+
+        Count count = (Count) o;
+
+        return compareTo(count) == 0;
+    }
+
+    @Override
+    public int hashCode()
+    {
+        return Long.hashCode(getValue(Magnitude.SINGLE));
+    }
+
+    public enum Magnitude
+    {
+        // must be in increasing magnitude order
+        SINGLE(1L, ""),
+        THOUSAND(1000L, "K"),
+        MILLION(1000_000L, "M"),
+        BILLION(1000_000_000L, "B"),
+        TRILION(1000_000_000_000L, "T"),
+        QUADRILLION(1000_000_000_000_000L, "P");
```

Comment:
This is not consistent with using `B` for Billion. How about `Q`?
```suggestion
        QUADRILLION(1000_000_000_000_000L, "Q");
```